lib/elixir/lib/enum.ex

```diff
@@ -1542,18 +1542,25 @@ defmodule Enum do
   ## Examples
 
       # Although not necessary, let's seed the random algorithm
-      iex> :rand.seed(:exsplus, {1, 2, 3})
-      iex> Enum.random([1, 2, 3])
-      2
+      iex> :rand.seed(:exsplus, {101, 102, 103})
       iex> Enum.random([1, 2, 3])
       1
+      iex> Enum.random([1, 2, 3])
+      3
 
   """
   @spec random(t) :: element | no_return
+  def random(enumerable)
+
+  def random(first..last),
+    do: random_integer(first, last)
+
   def random(enumerable) do
-    case take_random(enumerable, 1) do
-      [] -> raise Enum.EmptyError
-      [e] -> e
+    case count(enumerable) do
+      0 ->
+        raise Enum.EmptyError
+      count ->
+        at(enumerable, random_integer(0, count - 1))
```

Comment:
yes, it is faster in every single instance (i have excluded ranges since they are not part of the code in question),
bench file: https://gist.github.com/eksperimental/d9d98c1f4e47ecbcb19a6affe872dff4

```
$ mix bench -d 3
Settings:
  duration:      3.0 s

## EnumRandomBench
benchmark name                         iterations   average time 
Elixir.EnumRandomNew: list_empty         10000000   0.42 µs/op
Elixir.EnumRandomNew: map_empty          10000000   0.51 µs/op
Elixir.EnumRandomOriginal: list_empty    10000000   0.63 µs/op
Elixir.EnumRandomOriginal: map_empty      5000000   1.08 µs/op
Elixir.EnumRandomNew: list                1000000   8.03 µs/op
Elixir.EnumRandomNew: map                  100000   62.46 µs/op
Elixir.EnumRandomOriginal: list              5000   1030.92 µs/op
Elixir.EnumRandomOriginal: map               5000   1079.91 µs/op
Elixir.EnumRandomNew: list_big               5000   1342.06 µs/op
Elixir.EnumRandomNew: map_big                 500   13701.73 µs/op
Elixir.EnumRandomOriginal: list_big            50   223223.72 µs/op
Elixir.EnumRandomOriginal: map_big             50   251013.74 µs/op
``` 

see the results for the big maps and lists, the difference is HUGE, and the they have 200,000 elements